# Classwork 4: Data cleaning through Filtering

In this activity, you will clean a dataset using various techniques you learned in class for filtering and eliminating duplicate rows.

This activity is more free-form than the last ones. Its purpose is to get you to employ your critical thinking skills to form hypotheses and test them using the tools we've discussed in class until now. You'll need to interpret data and inspect closely, which are key skills in real projects with real datasets (like this one).


## Submitting this activity:
1. Complete the required challenges and questions
2. Download the file as a Jupyter Notebook by clicking File -> Download -> Download .ipynb
3. Rename the file into the following format: YOURID_Classwork4.ipynb
4. Upload the file into the appropriate place in Canvas

Run the cell below at least once. We won't have an autograder, but we need to load the function that loads the dataset.

In [ ]:
!git clone https://github.com/GzzDany/ADIAclasswork.git

We'll use the same dataset we worked with in the last activity. We'll assume all data conversion steps were properly done.

In [69]:
import pandas as pd
from ADIAclasswork.autograde_act4 import *


df = load_dataset()


# Mostrar los datos
df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee
0,2.0,2024-01-01 00:57:55,2024-01-01 01:17:43,1.0,1.72,1.0,N,186.0,79.0,2.0,17.7,1.0,0.5,0.00,0.0,1.0,22.70,2.5,0.0
1,1.0,2024-01-01 00:03:00,2024-01-01 00:09:36,1.0,1.80,1.0,N,140.0,236.0,1.0,10.0,3.5,0.5,3.75,0.0,1.0,18.75,2.5,0.0
2,1.0,2024-01-01 00:17:06,2024-01-01 00:35:01,1.0,4.70,1.0,N,236.0,79.0,1.0,23.3,3.5,0.5,3.00,0.0,1.0,31.30,2.5,0.0
3,1.0,2024-01-01 00:36:38,2024-01-01 00:44:56,1.0,1.40,1.0,N,79.0,211.0,1.0,10.0,3.5,0.5,2.00,0.0,1.0,17.00,2.5,0.0
4,1.0,2024-01-01 00:46:51,2024-01-01 00:52:57,1.0,0.80,1.0,N,211.0,148.0,1.0,7.9,3.5,0.5,3.20,0.0,1.0,16.10,2.5,0.0


First, we need to check how many values are missing from our dataset, and how dire the situation can be in terms of the proportion of our dataset with missing values.


Let's run the cell below to analyze this:



In [70]:
df.shape, df.isna().sum()

((2964624, 19),
 VendorID                      0
 tpep_pickup_datetime          0
 tpep_dropoff_datetime         0
 passenger_count          140162
 trip_distance                 0
 RatecodeID               140162
 store_and_fwd_flag       140162
 PULocationID                  0
 DOLocationID                  0
 payment_type                  0
 fare_amount                   0
 extra                         0
 mta_tax                       0
 tip_amount                    0
 tolls_amount                  0
 improvement_surcharge         0
 total_amount                  0
 congestion_surcharge     140162
 Airport_fee              140162
 dtype: int64)

Seeing as all the columns that have missing values have the same number of missing values, and they're not a significant proportion of the dataset, given that there's almost 3 million data points and only 140,162 missing values, we may consider removing them as the most viable option.

Hopefully there's significant overlap between columns with missing values, which would be the best case scenario.

We'll generate a new variable with the clean dataset to preserve the original dataset in memory in case we make some mistake in our analysis.

In [71]:
df_clean = df.dropna()
df_clean.shape

(2824462, 19)

Given the new amount of rows, our previous hypothesis was correct: there was complete overlap between the columns with missing values.

We'll work with df_clean from now on.

It's now time to think of possible errors that could be found in the data using context for what each column represents (for example, can some columns be negative? Are there upper or lower limits they should respect? etc).

In the following cells, you should validate whether the errors you are testing are indeed present in the dataset and if so, remove them. You must verify at least 3 possible errors and explain briefly what you are testing, how you are testing for those errors and what was the result of your process.

If you're out of ideas, you can go back to the previous activity and observe carefully what we explored within it... There are a lot of data errors and inconsistencies in the challenges.


Feel free to add cells with text and code as necessary, but keep your work clean and organized, and make sure it's easy to understand when you upload it.

1. In the previous activity we saw that the "fare_amount" column minimum was negative, which is an error. We will analyze how many values in that column are negative or 0 and remove them if possible.

In [72]:
### YOUR CODE HERE
mask = df_clean["fare_amount"]<= 0
mask.sum()


np.int64(36225)

Since there are 36,225 values with an error in this column and we have over 2.8 million data points, we can probably remove the error without a significant impact on our future analysis.

In [73]:
### YOUR CODE HERE

mask = df_clean["fare_amount"]>0

df_clean = df_clean[mask]

df_clean.shape



(2788237, 19)

The shape should now be (2,788,237, 19)

It's your turn now! Verify at least 3 characteristics about the dataset and if you find any inconsistencies in them, remove them! Please describe in text what you are trying to do in short, concise sentences.

In [ ]:
mask = df_clean["mta_tax"]< 0
mask.sum()
#1. This tells me there are tax values of 0, which there shouldn't be

np.int64(0)

In [33]:
mask = df_clean["mta_tax"]>=0

df_clean = df_clean[mask]

df_clean.shape
#With this, I removed all the 0 tax values

(2760913, 19)

In [52]:
mask = df_clean["passenger_count"]<= 0
mask.sum()
#2. This tells me there are passenger count values of 0 or negative, which there shouldn't be

np.int64(0)

In [53]:
mask = df_clean["passenger_count"]> 0

df_clean = df_clean[mask]

df_clean.shape
#With this, I removed all the 0 and negative passenger values

(2729788, 19)

In [74]:
mask = df_clean["trip_distance"]<= 0
mask.sum()
#3. This tells me there are trip distance values of 0, which there shouldn't be

np.int64(33772)

In [75]:
mask = df_clean["trip_distance"]> 0

df_clean = df_clean[mask]

df_clean.shape
#With this, I removed all the 0 and negative trip distance values

(2754465, 19)